jdaviz/configs/default/plugins/virtual_observatory/vo_plugin.py

```diff
@@ -0,0 +1,419 @@
+    resources = List([]).tag(sync=True)
+    resource_selected = Any().tag(sync=True)  # Any to accept Nonetype
+    resources_loading = Bool(False).tag(sync=True)
+
+    source = Unicode("").tag(sync=True)
+    coord_follow_viewer_pan = Bool(False).tag(sync=True)
+    viewer_centered = Bool(False).tag(sync=True)
+    coordframes = List([]).tag(sync=True)
+    coordframe_selected = Unicode("icrs").tag(sync=True)
+    radius_val = Float(1).tag(sync=True)
+    radius_unit_items = List().tag(sync=True)
+    radius_unit_selected = Unicode("deg").tag(sync=True)
+
+    results_loading = Bool(False).tag(sync=True)
+    data_loading = Bool(False).tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.viewer = ViewerSelect(
+            self, "viewer_items", "viewer_selected", manual_options=["Manual"]
+        )
+
+        self.radius_unit = UnitSelectPluginComponent(
+            self, items="radius_unit_items", selected="radius_unit_selected"
+        )
+        self.radius_unit.choices = ["deg", "rad", "arcmin", "arcsec"]
+
+        # Waveband properties to filter available registry resources
+        self.wavebands = [
+            w.lower() for w in vocabularies.get_vocabulary("messenger")["terms"]
+        ]
+        self.waveband_selected = None
+
+        self._full_registry_results = None
+        self.resource_selected = None
+
+        self.coordframes = [frame.lower() for frame in all_astropy_frames]
+
+        self.table.headers_avail = ["Title", "Instrument", "DateObs", "URL"]
+        self.table.headers_visible = ["Title", "Instrument", "DateObs"]
+        self._populate_url_only = False
+
+        self.table.show_rowselect = True
+        self.table.item_key = "URL"
+
+        self.hub.subscribe(self, AddDataMessage, handler=self.vue_center_on_data)
+        self.hub.subscribe(self, RemoveDataMessage, handler=self.vue_center_on_data)
+        self.hub.subscribe(self, LinkUpdatedMessage, handler=self.vue_center_on_data)
+
+    @observe("viewer_selected", type="change")
+    def vue_viewer_changed(self, _=None):
+        # Check mixin object initialized
+        if hasattr(self, "viewer"):
+            # Clear all existing subscriptions and resubscribe to selected viewer
+            for viewer in self.viewer.viewers:
+                if viewer == self.viewer.selected_obj:
+                    viewer.state.add_callback("zoom_center_x", self.vue_center_on_data)
+                    viewer.state.add_callback("zoom_center_y", self.vue_center_on_data)
+                else:
+                    # If not subscribed anyways, remove_callback should produce a no-op
+                    viewer.state.remove_callback(
+                        "zoom_center_x", self.vue_center_on_data
+                    )
+                    viewer.state.remove_callback(
+                        "zoom_center_y", self.vue_center_on_data
+                    )
```

Comment:
could this just use the new/old information in the event instead of looping through everything?

Reply:
I spent a LOONG time along this exact thread; basically I couldn't find a way to deterministically know "what was the viewer before the event call was made" without manually caching the `old_viewer` in a new attribute or something, which felt like overkill. I even looked at the viewer event to see if there was a way to change the even to send BOTH the viewer prior and the viewer after; but I came to the determination by the time the event is fired, the prior selection has been lost to time.

This is also considering the for loop is only over the viewers, which realistically I don't think we have an issue with having so many viewers that it will bog this down; the tool will slow down well in advance of this loop chugging, I'd assume